module/state_synchronization/requester/execution_data_requester.go

```diff
@@ -0,0 +1,640 @@
+	}
+
+	logger.Debug().Msgf("Fetched execution data for block %d", request.Height)
+
+	request.ExecutionData = executionData
+
+	e.status.Fetched(request)
+
+	return nil
+}
+
+// fetchExecutionData fetches the ExecutionData by its ID, and times out if fetchTimeout is exceeded
+func (e *executionDataRequester) fetchExecutionData(signalerCtx irrecoverable.SignalerContext, executionDataID flow.Identifier, height uint64) (executionData *state_synchronization.ExecutionData, err error) {
+	ctx, cancel := context.WithTimeout(signalerCtx, e.config.FetchTimeout)
+	defer cancel()
+
+	start := time.Now()
+	e.metrics.ExecutionDataFetchStarted()
+	defer func() {
+		// needs to be run inside a closure so the variables are resolved when the defer is executed
+		e.metrics.ExecutionDataFetchFinished(time.Since(start), err == nil, height)
+	}()
+
+	// Get the data from the network
+	// this is a blocking call, won't be unblocked until either hitting error (including timeout) or
+	// the data is received
+	executionData, err = e.eds.Get(ctx, executionDataID)
+
+	if err != nil {
+		return nil, err
+	}
+
+	// Write it to the local blobstore
+	_, _, err = e.eds.Add(signalerCtx, executionData)
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to write execution data to blobstore: %w", err)
+	}
+
+	return executionData, nil
+}
+
+// Notification Worker Methods
+
+func (e *executionDataRequester) processNotificationJob(ctx irrecoverable.SignalerContext, job module.Job, complete func()) {
+	// convert job into a block entry
+	entry, err := status.JobToBlockEntry(job)
+	if err != nil {
+		ctx.Throw(fmt.Errorf("failed to convert job to entry: %w", err))
+	}
+
+	e.processNotification(ctx, entry)
+	complete()
+}
+
+func (e *executionDataRequester) processNotification(ctx irrecoverable.SignalerContext, entry *status.BlockEntry) {
+	e.log.Debug().Msgf("notifying for block %d", entry.Height)
+
+	// ExecutionData may have been purged from the cache, look it up again
+	if entry.ExecutionData == nil {
+		e.log.Debug().Msgf("execution data not in cache for block %d", entry.Height)
+
+		result, err := e.results.ByBlockID(entry.BlockID)
+		if err != nil {
+			ctx.Throw(fmt.Errorf("failed to lookup execution result for block: %w", err))
+		}
+
+		entry.ExecutionData, err = e.eds.Get(ctx, result.ExecutionDataID)
+
+		if err != nil {
+			// At this point the data has been downloaded and validated, so it should be available
+			ctx.Throw(fmt.Errorf("failed to get execution data for block: %w", err))
+		}
+	}
+
+	// send notifications
+	e.notifyConsumers(entry.ExecutionData)
+	e.metrics.NotificationSent(entry.Height)
+}
+
+func (e *executionDataRequester) notifyConsumers(executionData *state_synchronization.ExecutionData) {
+
+	e.consumerMu.RLock()
+	defer e.consumerMu.RUnlock()
+
+	for _, fn := range e.consumers {
+		fn(executionData)
+	}
+}
+
+// Bootstrap Methods
+
+func (e *executionDataRequester) checkDatastore(ctx irrecoverable.SignalerContext, lastHeight uint64) error {
```

Comment:
This happens every restart when startup checks are enabled. making this stateful is a good idea especially if a node is crashing. maybe it could be a job queue, and when it reaches the end, the consumer is disabled, and the `processedIndex` is reset